pulsar-broker/src/main/java/org/apache/pulsar/broker/service/PulsarCommandSenderImpl.java

```diff
@@ -263,17 +268,22 @@ public ChannelPromise sendMessagesToConsumer(long consumerId, String topicName,
 
                 int redeliveryCount = redeliveryTracker
                         .getRedeliveryCount(PositionImpl.get(entry.getLedgerId(), entry.getEntryId()));
-
                 ctx.write(
                         cnx.newMessageAndIntercept(consumerId, entry.getLedgerId(), entry.getEntryId(), partitionIdx,
                                 redeliveryCount, metadataAndPayload,
                                 batchIndexesAcks == null ? null : batchIndexesAcks.getAckSet(i), topicName, epoch),
                         ctx.voidPromise());
-                entry.release();
+                entriesToRelease.add(entry);
             }
 
             // Use an empty write here so that we can just tie the flush with the write promise for last entry
             ctx.writeAndFlush(Unpooled.EMPTY_BUFFER, writePromise);
+            writePromise.addListener((future) -> {
+                // release the entries only after the flush
+                // otherwise we may think that the memory has been released, but that's not correct
+                // this is happening on the same thread as the writes above
```

Comment:
I got confused with the wording. It might be helpful to rephrase it and explain the relation to memory limits etc.